llvm/lib/Analysis/ValueTracking.cpp

```diff
@@ -868,10 +873,9 @@ void llvm::computeKnownBitsFromContext(const Value *V, KnownBits &Known,
 
     Value *Arg = I->getArgOperand(0);
 
-    if (Arg == V && isValidAssumeForContext(I, Q.CxtI, Q.DT)) {
-      assert(BitWidth == 1 && "assume operand is not i1?");
-      (void)BitWidth;
-      Known.setAllOnes();
+    if (match(Arg, m_TruncOrSelf(m_Specific(V))) &&
+        isValidAssumeForContext(I, Q.CxtI, Q.DT)) {
+      Known.One.setBit(0);
```

Comment:
You can `sext`/`zext` if you have `nsw`/`nuw` on the trunc: https://alive2.llvm.org/ce/z/dx38fL

Reply:
maybe better to call `computeKnownBitsFromCond` instead after https://github.com/llvm/llvm-project/pull/126414 so it is handled in one place

Reply:
have added handling of nuw but avoided the nsw for same reason as here https://github.com/llvm/llvm-project/pull/126414#discussion_r1948159697